### PR TITLE
Refactor Trie hasher using onRoot flag

### PR DESCRIPTION
**Pull request by blukat29**
## Proposed changes

De-duplicate code in the trie hasher (storage/statedb/hasher.go).

- Current implementation has two pairs of similar functions:
  - single-threaded version `hash()` + `hashChildren()`
  - multi-threaded version `hashRoot()` + `hashChildrenFromRoot()`
- This PR removes the latter. Instead, the `onRoot` argument switches multi-threading.
  - `hashNode(..., onRoot)` + `hashChildren(..., onRoot)` runs multi-threaded if onRoot == true.
	  - `hash(...) = hashNode(..., onRoot=false)`
	  - `hashRoot(...) = hashNode(..., onRoot=true)`

## Types of changes

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

Benchmark results show that this PR does not affect hashing performance (on Apple M1 Max 10-core)
- Before PR
	```
	$ go test -run=^$ -bench=BenchmarkHash
	BenchmarkHash-10                 2582688               402.2 ns/op           684 B/op          7 allocs/op
	BenchmarkHashFixedSize/10-10       26340             43323 ns/op           14891 B/op        179 allocs/op
	BenchmarkHashFixedSize/100-10              12986             89688 ns/op           66881 B/op        809 allocs/op
	BenchmarkHashFixedSize/1K-10                2024            618712 ns/op          659172 B/op       7811 allocs/op
	BenchmarkHashFixedSize/10K-10                282           4171724 ns/op         6907710 B/op      79703 allocs/op
	BenchmarkHashFixedSize/100K-10                33          35102691 ns/op        68240330 B/op     794537 allocs/op
	```
- After PR
	```
	BenchmarkHash-10                 3081148               439.9 ns/op           677 B/op          7 allocs/op
	BenchmarkHashFixedSize/10-10       27524             42308 ns/op           14262 B/op        178 allocs/op
	BenchmarkHashFixedSize/100-10              13624             88954 ns/op           66293 B/op        808 allocs/op
	BenchmarkHashFixedSize/1K-10                2050            604276 ns/op          658532 B/op       7810 allocs/op
	BenchmarkHashFixedSize/10K-10                280           4168147 ns/op         6906603 B/op      79701 allocs/op
	BenchmarkHashFixedSize/100K-10                30          38524168 ns/op        68240050 B/op     794539 allocs/op
	```
- (for comparison) After PR, if onRoot == false
	```
	BenchmarkHash-10                  564274              2035 ns/op             628 B/op          7 allocs/op
	BenchmarkHashFixedSize/10-10       41187             28873 ns/op           13213 B/op        155 allocs/op
	BenchmarkHashFixedSize/100-10               8718            134050 ns/op           64321 B/op        772 allocs/op
	BenchmarkHashFixedSize/1K-10                 842           1415162 ns/op          649540 B/op       7742 allocs/op
	BenchmarkHashFixedSize/10K-10                 70          17122263 ns/op         6865380 B/op      79500 allocs/op
	BenchmarkHashFixedSize/100K-10                 7         164345345 ns/op        68196154 B/op     794339 allocs/op	
	```

### CODE REVIEW

File: storage/statedb/hasher.go [hunk 1]
```diff
@@ -74,48 +74,14 @@ func returnHasherToPool(h *hasher) {
...
-	return hashed, cached
+// hashRoot is similar to hash() but adds special treatment for the root node
+func (h *hasher) hashRoot(n node, db *Database, force bool) (node, node) {
+	return h.hash(n, db, force, true)
 }
 
-func (h *hasher) hashRoot(n node, db *Database, force bool) (node, node) {
+// hash collapses a node down into a hash node, also returning a copy of the
+// original node initialized with the computed hash to replace the original one.
+func (h *hasher) hash(n node, db *Database, force bool, onRoot bool) (node, node) {
```
Comment (ethan-kr) [inline storage/statedb/hasher.go:84]: how about this line change to

func (h *hasher) hash(n node, db *Database) (node, node) {
    return hashProc(, db, force, false)
}

func  (h *hasher) hashProc(n node, db *Database, force bool) (node, node) {

**Review by ethan-kr — APPROVED** [review]
I made a comment, but there is no disagreement, so I approve it.

File: storage/statedb/hasher.go (hunk 1 above)
Comment (blukat29) [reply storage/statedb/hasher.go:84]: Okay. I will call the internal function, say `hashNode`.
PTAL when I push another commit.

**Review by jeongkyun-oh — APPROVED** [review]
LGTM

**Comment by blukat29** [comment]
@ethan-kr PTAL :)

**Review by aidan-kwon — APPROVED** [review]

**Review by ethan-kr — APPROVED** [review]

**Pull request merged**